as-ps/as_ps/main.py

```diff
@@ -100,64 +110,99 @@ def write(self, reason, value):
-                self.driver.setParam(reason, new_value)
-            self.driver.setParamStatus(
-                reason, _Alarm.NO_ALARM, _Severity.NO_ALARM)
+    def _write_operation(self, bbb, pvname, value):
+        t0 = _time.time()
+        bbb.write(pvname.device_name, pvname.propty, value)
+        # NOTE: This scan_device might be redundent
+        self._scan_device(bbb, pvname.device_name, force_update=True)
+        t1 = _time.time()
+        _log.info("[{:.2s}] - {:.32s} = {:.50s}".format(
+            'T ', pvname, 'write operation took {:.3f} ms'.format((t1-t0)*1000)))
 
-    def _update_ioc_database_disconnected(self, bbb, device_name):
+    def _scan_bbb(self, bbb):
+        for device_name in bbb.psnames:
+            self._scan_device(bbb, device_name)
+
+    def _scan_device(self, bbb, device_name, force_update=False):
+        bbb_connected = bbb.check_connected(device_name)
+        self._update_ioc_database(bbb, device_name, bbb_connected, force_update)
+
+    def _update_ioc_database(self, bbb, device_name,
+                             bbb_connected=True,
+                             force_update=False):
         # Return dict indexed with reason
-        data, updated = bbb.read(device_name)
+        data, updated = bbb.read(device_name, force_update=force_update)
+        # data, updated = self._bbb_read(bbb, device_name,
+        #                                force_update=force_update)
         if not updated:
             return
+
         for reason, new_value in data.items():
+            if self._prucqueue and App._setpoint_regexp.match(reason):
+                # While there are pending write operations in the queue we
+                # cannot update setpoint variables or we will spoil the
+                # accepted value in the write method.
+                continue
             if self._check_value_changed(reason, new_value):
                 if new_value is None:
                     continue
                 self.driver.setParam(reason, new_value)
-            self.driver.setParamStatus(
-                reason, _Alarm.TIMEOUT_ALARM, _Severity.INVALID_ALARM)
-
-    def _scan_bbb(self, bbb):
-        for device_name in bbb.psnames:
-            if bbb.check_connected(device_name):
-                self._update_ioc_database(bbb, device_name)
+            if bbb_connected:
+                self.driver.setParamStatus(
+                    reason, _Alarm.NO_ALARM, _Severity.NO_ALARM)
             else:
-                self._update_ioc_database_disconnected(bbb, device_name)
+                self.driver.setParamStatus(
+                    reason, _Alarm.TIMEOUT_ALARM, _Severity.INVALID_ALARM)
+
+    def _check_value_changed(self, reason, new_value):
+        old_value = self.driver.getParam(reason)
+        if isinstance(new_value, _np.ndarray):
+            # TODO: improve here!
+            return True
+        else:
+            # TODO: a python-list of 4000 elements takes ~70us for
```

Comment:
If I'm not wrong, pcaspy works with numpy arrays when dealing with waveform channels, so it is not necessary to convert the value passed to the write method to numpy arrays.

Reply:
I understand that, bu `bbb.read` gets `WfmData-SP/RB` from class objects that manipulate these parameters as python list. What I meant in the comment was that we should eventually convert to numpy arrays from the bottom (classes used before updating epics db).